### ISSUE TITLE
Process Porto data

### ISSUE THREAD

**Issue by BrunoRosendo** [BrunoRosendo/master-thesis#92]
After #91, I should process the data retrieved from Porto's metro and bus in order to properly run my algorithms. This might include organizing the available tables, using Google Maps if the time/distance between stations is not available or looking for extra data.

I think the trickiest part will be RPP, since I don't have data on the passengers' trips. I may look for this information elsewhere or try to come up with an assumption

**Comment by BrunoRosendo** [comment]
After analysing the metro's data, it only makes sense to use RPP since metros always go back and forth instead of being in a circle.

I reached the following steps:
- [ ] We have good information about the geographic places of the stops, which almost directly gives us the locations for the algorithm.
- [ ] As for the distance matrix, we can use the stop times to determine the time between stops.
- [ ] The number of vehicles can be considered the number of lines, and I should change this value to compare.
- [ ] I need to look for the capacity of the metro; that should be easy.
- [ ] The worst part is the trips since the data doesn't contain passengers' information. The simplest way is to consider every stop in the line as a trip and adjust the capacity accordingly. Otherwise, I need more data or another sort of assumption.
- [ ] At the end, I can try to put the map of Porto in the background

**Comment by BrunoRosendo** [comment]
As for STCP buses:
- [ ] There are circular lines, so I can use those as an example of a single bus line with CVRP.
- [ ] For multiple lines together, I can do the same steps as for the metro. There are A LOT of stops, so choose a few lines wisely.

**Comment by BrunoRosendo** [comment]
For some reason, RPP is not respecting the "no return to depot" penalty and is ignoring a few stops in the end... I need to figure this out

**Comment by BrunoRosendo** [comment]
The classic solver is also having some trouble. The stops do not correspond with the line I'm testing

**Comment by BrunoRosendo** [comment]
> For some reason, RPP is not respecting the "no return to depot" penalty and is ignoring a few stops in the end... I need to figure this out. I could simply give the algorithm more time, I need to test this
> 
> I can also check if it's related to the normalization and the really big numbers in the distance matrix, but I don't think so...
> 
> Edit: I think the problem is just too big, I need to try with fewer vars

I tried running in main.py after copying the simplified input from the classic solver and I got the right result... I will check if there's something wrong with how the RPP algorithm simplifies the input

**Comment by BrunoRosendo** [comment]
> > For some reason, RPP is not respecting the "no return to depot" penalty and is ignoring a few stops in the end... I need to figure this out. I could simply give the algorithm more time, I need to test this
> > I can also check if it's related to the normalization and the really big numbers in the distance matrix, but I don't think so...
> > Edit: I think the problem is just too big, I need to try with fewer vars
> 
> I tried running in main.py after copying the simplified input from the classic solver and I got the right result... I will check if there's something wrong with how the RPP algorithm simplifies the input

Conclusion: Everything is fine with the algorithm, it was just the stop trindade being in a different place in the locations array, even though it doesn't make a difference for the problem definition. However, the optimizer seemed to randomly get to the right solution sooner when using one of the formulations (the simplified one)

**Comment by BrunoRosendo** [comment]
Here's the RPP algorithm running on the D line stops (not a surprising result, since the line is straightforward)

![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/e3fb8db5-0040-47ed-91e7-7e7f6aea89c0)

**Comment by BrunoRosendo** [comment]
This is line A, with 1000+ vars, running on dwave hybrid system. This is impressive, especially since the local CPLEX optimizer can't even run this on the free version

![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/67cb5f75-e23b-4695-b483-7ad5ac1d3e26)

**Comment by BrunoRosendo** [comment]
I'm having some trouble stretching the y axis, plotly is not making things easy...
A workaround is to set the height of the figure and then zoom out like this

![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/7b27e176-503e-4724-b645-3ef0c9b73b36)

**Comment by BrunoRosendo** [comment]
Ignoring previous comments, these are the tasks left:
- [ ] Fix the SELECTED_TRIP_COUNT logic, using 2 disjoint bus lines yielding strange results with it. Using all the stops is too complex.
- [ ] Circular bus lines with CVRP
- [ ] Multiple lines with shared stops (repeat locations?)
- [ ] Refactor #97

**Comment by BrunoRosendo** [comment]
![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/5f4d9f43-44ee-4108-88f4-bcff4c726776)

Circular line being handled by CVRP

**Comment by BrunoRosendo** [comment]
![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/c14982ff-aafa-49da-bca7-086732af418a)

Regular bus line

**Comment by BrunoRosendo** [comment]
![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/74a6c0ac-4466-42ed-8e73-756180370723)
Multiple bus lines are now working :partying_face: 
This was 30s on dwave hybrid solver

**Comment by BrunoRosendo** [comment]
![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/a61a5384-3d8f-48d4-8019-ca000d436193)

This was 3838 vars/qubits on dwave hybrid solver, in 30 seconds. UHUUU lines 18+304

**Comment by BrunoRosendo** [comment]
I implemented a solution for multiple lines but my first tests didn't go very well. I should test more and fix the problems

**Comment by BrunoRosendo** [comment]
After some more testing, I have new conclusions.

RPP:
- Single line :white_check_mark: 
- Multiple adjoint lines :white_check_mark: 
- Multiple lines w/ common stops :x:

This happens because each location is only visited once and, if we try duplicating them, it's impossible to satisfy all trip requests. I tried satisfying only some trips but that resulted in unexpected behaviour from the algorithms. OR tools doesn't even yield a solution.

CVRP:
- Single line :white_check_mark: 
- Multiple adjoint lines :x:
- Multiple lines w/ common stops :white_check_mark: 

Adjoint lines aren't possible because we need a common depot.
It's possible to support multiple lines if ALL of them have a common stop (the new depot). I need to duplicate other shared stops (w/ no connection between them) and make sure the dist. matrix is bi-directional, since we're changing the depot

**Comment by BrunoRosendo** [comment]
> ![image](https://private-user-images.githubusercontent.com/55768934/333689678-78bddc7d-3202-4105-843d-022be6547b1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1NzM5MjEsIm5iZiI6MTcxNjU3MzYyMSwicGF0aCI6Ii81NTc2ODkzNC8zMzM2ODk2NzgtNzhiZGRjN2QtMzIwMi00MTA1LTg0M2QtMDIyYmU2NTQ3YjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI0VDE4MDAyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjZjFkY2ZhNWUwZDliMDc1NzY4ZjFkNGJjOTQ2MjBkYTZlODdmMTg2ZTM2OWY4MDJmYmFjMmUyZWUzZGQxNzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HMrb3wtAXJ_qY7UzX0oUGEk7-ToiKdcNulEIwo-TARA) This are lines 300 and 3002 together with 2 common stops and 1 vehicle. The result is not surprising

As we can see, doing the same example with the distance API yields a different result. I think I'm finally able to get something interesting out of this

![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/fd2e63cf-f688-48aa-9cba-354c958f125f)

**Comment by BrunoRosendo** [comment]
> There's still something wrong, sometimes not all connections are possible and I get infeasible results

I think I solved this with the logic below, there were some shenanigans with when I should be duplicating the depot

```python
# Delete duplicates of the new depot. Needed to avoid unfeasible solutions with multiple vehicles.
    to_delete = min(NUM_VEHICLES - 1, most_common_freq - 1)
    i = 1
    while i < len(location_ids) and to_delete > 0:
        if location_ids[i] == most_common_id:
            del location_ids[i]
            del locations[i]
            del location_names[i]
            to_delete -= 1
        else:
            i += 1
```

Now I can work with 2 lines + 2 vehicles, for example.

![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/e576ca92-a1c5-4f56-a58d-390646f3a0be)

or even more vehicles, although the result is the same

![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/81f21314-84a7-4c1b-9765-980ab98f087a)

**Comment by BrunoRosendo** [comment]
I believe I can do useful things with my current setup. For example, 3 lines together:

![image](https://github.com/BrunoRosendo/master-thesis/assets/55768934/ce41600b-de94-4d34-9c36-222075de2e3a)

Or at least confirm if the current lines are optimized (order of stops, not client satisfaction).

I can also try to add a few stops if I want. I will create better examples while I'm writing the thesis

THIS CLOSES THE DEVELOPMENT PHASE UHUUU :partying_face:
